SQS Secure Transport

kylehuntsman opened the pull request:
Ticket: [RUI-1379](https://jira.meltdev.com/browse/RUI-1379)

Added the ability to send sensitive data through SQS via encryption. Sensitive data is handed to the SQS `sendMessage()` function, and it will encrypt it, add it to the message body, and send it to the queue. On the consumer side, it will take the message, decrypt it, and send it along like normal. The user should not have to worry about writing encryption code, other than setting up the encryption configuration values.

### Overview
- The original SQS `sendMessage()` function now takes an optional second argument for data that should be encrypted.
- Encrypted data is attached to the original message body as a child object on the `encrypted` key. This child object has a data property, which is the encrypted data, and a key property, which is an encrypted version of the key used to encrypt the data.
- Encryption is done via AWS KMS.
- A new `encryption` property was added to the config to allow customization of the KMS key name to use for encryption. The configuration values must be set for the data to be encrypted and sent along.
  - key: The name of the KMS key to generate a key from, which will be used for the encryption.
- If data is passed without the configuration set, an `EncryptionConfigurationError` is thrown.

```js
// Configuration
{
  awsext: {
    sqs: {
      my-consumer: {
        queue: { ... },
        sqsOptions: { ... },
        kmsOptions: { ... }, // <-- Options to be passed to the KMS instance
        encryption: { // <-- Encryption configuration for sending sensitive data through SQS
          key: 'Key for SQS secure transport' // <-- KMS master key name
        }
      }
    }
  }
}
```
```js
// SQS Message Body with Encrypted Data
{
  myProperty: 'myValue',
  encrypted: {
    data: 'Ab48EF303fbeAcd908tKj2erMnd3',
    key: 'Bc59FG414gcdBde019uLk3dsNoe4'
  }
}
```

kylehuntsman committed:
Removed support for S3 storage, and added support for SQS message encryption

kylehuntsman committed:
Added decryption to sqs-consumer

kylehuntsman committed:
Fixed test errors

kylehuntsman committed:
Updated encryption tests and sendMessage() promise logic

kylehuntsman committed:
Added an EncryptionConfigurationError

This error is thrown when sensitive data is passed to the sqs sendMessage(), but no encryption configuration was found.

kylehuntsman committed:
Now checking for imporper config in the consumer as well.

kylehuntsman committed:
Updated tests and consumer decryption promise logic

kylehuntsman committed:
Fixing travis failures

sukrit007 commented on lib/sqs/sqs-base.js:
```diff
@@ -89,19 +93,40 @@ class SqsBase extends events.EventEmitter {
+
+        // else just validate the normal message
+        return this.validateMessage(msgData)
+          .then(() => msgData);
```
Not sure why we need this line

sukrit007 replied:
My bad .  It is clear now. THanks

sukrit007 commented on lib/sqs/sqs-base.js:
```diff
@@ -5,11 +5,13 @@ require('promise.prototype.finally');
 const
   promisify = require("es6-promisify"),
   config = require("config"),
+  crypto = require('crypto'),
```
Not needed

sukrit007 commented on test/unit/common/encryption-spec.js:
```diff
@@ -0,0 +1,95 @@
+
+      it('has the correct encrypted data', () => {
+        return promise.then(payload => {
+          payload.data.should.eql(ENCRYPTED_DATA);
```
We should not compare encrypted data and possibly use a random salt for encryption

kylehuntsman committed:
Updated encryption to use AES256 by default

kylehuntsman committed:
Attempting to integrate with KMS. Running into some problems persisting the keys.

kylehuntsman committed:
Updated encryption integration with KMS

The encryption key buffers from KMS are being converted to Hex before usage and transport.

kylehuntsman committed:
Updated encryption logic to work with KMS

sukrit007 commented on examples/config/default.yaml:
```diff
@@ -4,13 +4,18 @@ awsext:
+        region: 'us-east-1'
+        # endpoint: http://localhost:9324
+      encryption:
+        key: 'alias/picard-sqs'
```
can you rename picard by example-sqs

sukrit007 commented on lib/common/encryption.js:
```diff
@@ -0,0 +1,88 @@
+        try {
+          return JSON.parse(dataString);
+        } catch(e) {
+          return dataString;
```
@FunnyGopher  any reason why we are returning data string and not throwing exception?

kylehuntsman replied:
This is incase the data being decrypted was instead a string instead of a json object

sukrit007 committed:
Change example key

sukrit007 reviewed (approved):
LGTM!

sukrit007 committed:
Comment out encryption by default ine xample
